_posts/contests/2021-02-05-Performance-analysis-and-tuning-contest-4.md

```diff
@@ -0,0 +1,104 @@
+---
+layout: post
+title: Performance analysis and tuning contest &#45&#51.
+categories: contest
+---
+
+**Contents:**
+* TOC
+{:toc}
+
+Welcome to the 4th edition of our performance analysis and tuning contest. If you see this post and haven't read my initial post about the [contest]({{ site.url }}/blog/2019/02/02/Performance-optimization-contest), I encourage you to read it first. 
+
+The fourth edition of the contest will be ran by Ivica Bogosavljevic from [Johny's Software Lab](https://johnysswlab.com) blog. Ivica also writes about software performance, so feel free to go and check out his blog, there is a ton of useful content there.
+
+------
+**Subscribe to my [mailing list](https://mailchi.mp/4eb73720aafe/easyperf), support me on [Patreon](https://www.patreon.com/dendibakh) or by PayPal [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=TBM3NW8TKTT34&currency_code=USD&source=url).**
+
+------
+
+Guest writer: [Ivica Bogosavljevic](https://johnysswlab.com/author/ibogi/)
+
+The benchmark for the 4th edition is canny edge detection algorithm. The source code, compilation and run script as well as test image is available in the [code repository](TODO: Add link to the repository).
+
+Canny is an image edge detector algorithm that exists for a long time. You can find more information about it in the [Wikipedia article](https://en.wikipedia.org/wiki/Canny_edge_detector), but they say an image is worth a thousand words, so here is before and after image so you can get the impression of how it works.
+
+TODO(insert image)
+
+### Quickstart
+
+
+The implementation used for this contest is available [online](http://www.math.tau.ac.il/~turkel/notes/canny_source.c). The same version (with very few changes) is available in the repository.
+
+To download and build canny do the following:
+```bash
+$ TODO(how to download a single directory from github?)
+$ cd canny_baseline
+$ mkdir build
+$ cd build
+$ cmake .. -DCMAKE_C_COMPILER=clang
+$ make
+```
+
+To run the benchmark:
+```bash
+$ ../run-and-compare.sh
+```
+
+If the program finished correctly, and the image it produced is good, you will see information about the runtime and a message `Validation successful`.
+
+You may also find useful Denis' python [script](https://github.com/dendibakh/dendibakh.github.io/tree/master/_posts/code/PerfContest/4/run.py) for conducting multiple experiments. See decription inside it.
+
+We use CLANG for this experiment since it produced faster binary in my tests.
+
+Target machine for this edition of the contest is again Haswell CPU with 64-bit Linux. Although you can do your experiments on Windows since `cmake` is used for building the benchmark. If you choose Windows as a platform, here is the article that might be helpful: [How to collect CPU performance counters on Windows?]({{ site.url }}/blog/2019/02/23/How-to-collect-performance-counters-on-Windows).
```

Comment:
I suggest targeting Skylake machines instead of Haswell. Let's also say that people are free to use any HW they want, be it AMD, Intel, or maybe even ARMs. 4x improvements should be visible on any decent HW.
But as a disclaimer, we will evaluate all submissions on a specific HW (I have access to most generations of Intel and AMD CPUs, so I can help measure them). In the end, it is not about getting the best score, but about practicing perf optimizations, so people should be fine with it.

Reply:
You will see large improvements on any hardware